store/sqlstore/integrity.go

```diff
@@ -0,0 +1,504 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package sqlstore
+
+import (
+	"fmt"
+
+	"github.com/mattermost/gorp"
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/store"
+)
+
+type relationalCheckConfig struct {
+	parentName         string
+	parentIdAttr       string
+	childName          string
+	childIdAttr        string
+	canParentIdBeEmpty bool
+	sortRecords        bool
+}
+
+func getOrphanedRecords(dbmap *gorp.DbMap, cfg relationalCheckConfig) ([]store.OrphanedRecord, error) {
+	var records []store.OrphanedRecord
+
+	query := fmt.Sprintf(`SELECT %s AS ParentId`, cfg.parentIdAttr)
+
+	if cfg.childIdAttr != "" {
+		query += fmt.Sprintf(` , %s AS ChildId`, cfg.childIdAttr)
```

Comment:
Propose we leverage github.com/Masterminds/squirrel for all query building vs. string concatenation -- you'll find various existing examples of this in the sqlstore package.

Reply:
I've actually tried using squirrel originally but thought the resulting code was harder to understand than what I eventually used (and it still included some string concatenations). That said, I am new to squirrel so maybe if I dig a bit more into its api I could come up with a better version of the query. I'll update on that soon.

Reply:
This should be an equivalent query built using squirrel:

```go	
        sub := sq.Select("TRUE").
		From(cfg.parentName).
		Prefix("NOT EXISTS (").
		Suffix(")").
		Where(sq.Eq{"id": cfg.childName + "." + cfg.parentIdAttr})

	main := sq.Select().
		Column(cfg.parentIdAttr + " AS ParentId").
	        From(cfg.childName).
	        Where(sub)

	if cfg.childIdAttr != "" {
		main = main.Column(cfg.childIdAttr + " AS ChildId")
	}

	if cfg.canParentIdBeEmpty {
		main = main.Where(sq.And{sq.NotEq{cfg.parentIdAttr: ""}})
	}

	if cfg.sortRecords {
		main = main.OrderBy(cfg.parentIdAttr)
	}

	query, args, _ := main.ToSql()
```

Not sure if readability can be improved much. I honestly find it to be more fragmented than the original plain text version but might be easier to extend in the future. Thoughts?

Reply:
I believe we've found it best to converge on squirrel whenever we find the need to conditionally build up SQL. It's perhaps not so obviously true in this case without the reuse of the expression and with the pain that is `Prefix`/`Suffix`. 2/5, but IMO it's still a net win.

A few tweaks that I think can make your version slightly clearer, e.g.:

This:
```go
		main = main.Where(sq.And{sq.NotEq{cfg.parentIdAttr: ""}})
```

can just be:
```go
		main = main.Where(sq.NotEq{cfg.parentIdAttr: ""})
```

Also I'd recommend indenting consistently, e.g.:
```go
 sub := sq.
        Select("TRUE").
		From(cfg.parentName).
		Prefix("NOT EXISTS (").
		Suffix(")").
		Where(sq.Eq{"id": cfg.childName + "." + cfg.parentIdAttr})
```